sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala

```diff
@@ -24,6 +24,24 @@ import org.apache.spark.sql.catalyst.plans._
 import org.apache.spark.sql.catalyst.plans.logical._
 
 /**
+ * A pattern that matches any number of project if fields is deterministic
+ * or child is LeafNode of project on top of another relational operator.
+ */
+object ProjectOperation extends PredicateHelper {
+  type ReturnType = (Seq[NamedExpression], LogicalPlan)
+
+  def unapply(plan: LogicalPlan): Option[ReturnType] = plan match {
+    case Project(fields, child) if fields.forall(_.deterministic) =>
+      Some((fields, child))
+
+    case Project(fields, child: LeafNode) if !fields.forall(_.deterministic) =>
```

Comment:
aha,  :(  
I misunderstood you.
Is not that we have added a condition.
`case p @ Project(fields, child: LeafNode) if p.references.nonEmpty =>`